torchrl/objectives/costs/dreamer.py

```diff
@@ -0,0 +1,175 @@
+# LICENSE file in the root directory of this source tree.
+
+
+from turtle import forward
+
+import torch
+import torch.distributions as d
+import torch.nn as nn
+from torch.distributions import kl_divergence
+
+from torchrl.data import TensorDict
+from torchrl.objectives.costs.common import LossModule
+from torchrl.objectives.costs.utils import LogLikelihood
+from torchrl.objectives.costs.utils import hold_out_net
+from torchrl.objectives.returns.functional import vec_td_lambda_return_estimate
+
+
+class DreamerModelLoss(LossModule):
+    """
+    Dreamer Model Loss
+    Computes the loss of the dreamer model given a tensordict that contains, prior and posterior means and stds,
+    the observation, the reconstruction observation, the reward and the predicted reward.
+
+    Args:
+        reco_loss (nn.Module): loss function for the reconstruction of the observation
+        reward_loss (nn.Module): loss function for the prediction of the reward
+
+    """
+
+    def __init__(
+        self,
+        world_model: nn.Module,
+        cfg: "DictConfig",
+        lambda_kl: float = 1.0,
+        lambda_reco: float = 1.0,
+        lambda_reward: float = 1.0,
+        reco_loss: nn.Module = LogLikelihood(reduction="none"),
+        reward_loss: nn.Module = LogLikelihood(),
+        free_nats: int = 3,
+    ):
+        super().__init__()
+        self.world_model = world_model
+        self.cfg = cfg
+        self.reco_loss = reco_loss
+        self.reward_loss = reward_loss
+        self.lambda_kl = lambda_kl
+        self.lambda_reco = lambda_reco
+        self.lambda_reward = lambda_reward
+        self.free_nats = free_nats
+
+    def forward(self, tensordict: TensorDict) -> torch.Tensor:
+        tensordict.batch_size = [tensordict.shape[0]]
+        tensordict["prior_state"] = torch.zeros(
+            (tensordict.batch_size[0], self.cfg.state_dim)
+        )
+        tensordict["belief"] = torch.zeros(
+            (tensordict.batch_size[0], self.cfg.rssm_hidden_dim)
+        )
+        tensordict = self.world_model(tensordict)
+        # compute model loss
+        kl_loss = self.kl_loss(
+            tensordict["prior_means"],
+            tensordict["prior_stds"],
+            tensordict["posterior_means"],
+            tensordict["posterior_stds"],
+        )
+        reco_loss = (
+            0.5
+            * self.reco_loss(
+                tensordict.get("pixels"),
+                tensordict.get("reco_pixels"),
+            )
+            .mean(dim=[0, 1])
+            .sum()
+        )
+        reward_loss = 0.5 * self.reward_loss(
+            tensordict.get("reward"), tensordict.get("predicted_reward")
+        )
+        loss = (
+            self.lambda_kl * kl_loss
+            + self.lambda_reco * reco_loss
+            + self.lambda_reward * reward_loss
+        )
+        return (
+            TensorDict(
+                {
+                    "loss_world_model": loss,
```

Comment:
Here you are returning 4 losses, one of them being a sum of the others. To make sure that we don't do anything stupid like re-summing the losses (which is what the trainer will do) you should either return one loss only (but then we won't be able to log each of them individually) or return the decomposed loss only.

Reply:
From what i 've seen from trainer it doesnt retrieve losses if it doesnt start with "loss_". So renaming loss_kl by kl and so on would do the trick no?

Reply:
yeah but you sum it above!